core/src/main/java/com/graphhopper/storage/BaseGraph.java

```diff
@@ -1201,12 +1154,9 @@ public final EdgeIteratorState detach(boolean reverseArg) {
         }
     }
 
-    /**
-     * Common private super class for AllEdgesIteratorImpl and EdgeIterable
-     */
-    static abstract class CommonEdgeIterator implements EdgeIteratorState {
+    static class EdgeIteratorStateImpl implements EdgeIteratorState {
```

Comment:
Renamed this because its more like an 'edge' than an iterator. Its used as base class for the iterators yes, but there wasn't much that was 'iterator-like' about this class.

Reply:
Thanks! I think it is probably time to try again if we can separate the EdgeIterator interface from EdgeIteratorState (the last time I tried this it was still to heavily coupled in CH but maybe now this is possible?)

Reply:
Ok, but why would this be better? An EdgeIterator is just an EdgeIteratorState with a `next()` method. I don't think this ever bothered me.

Reply:
Hmmh, my argument would be that they should be separated like Object is separate from Iterator, but yes, here we change the properties of "Object" with every `iterator.next()` call and so `EdgeIterator` is not really an Iterator it is more an EdgeIteratorState with a `setPointer/setCursor` method (very similar to the [ResultSet](https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/ResultSet.html) from jdbc API)

Reply:
Ok for me an `EdgeIteratorState` is a proxy to a row of data, or in some cases its an actual row of data (after calling `detach()`). An `EdgeIterator` is the same thing except that it can be moved to another edge. I think @michaz suggested calling it a 'cursor' because it changes state after calling `next()`, compared to the `java.util.Iterator` which *returns* the next element when calling `next()`. But honestly, I am so used to it I do not really care whether its called iterator or cursor. But I do not see real value in separating `EdgeIteratorState` from `EdgeIterator`, because like I said in this case the latter is really the same thing as the first except that it allows to be shifted (using the `next()` method).

Now we are at it how would we call an `EdgeIteratorState` that can be set to an arbitrary edge (not just move to the next edge at the same base node). Would it be a `SingleEdgeExplorer` or a `SingleEdgeIterator` or a `SingleEdgeCursor`? :smiley:

Reply:
Probably the EdgeIterator is not a "cursor" because it has no direct setEdgeId method, only an implicit "next()" and so "iterator" is more appropriate (?)

> Now we are at it how would we call an EdgeIteratorState that can be set to an arbitrary edge (not just move to the next edge at the same base node).

I like SingleEdgeCursor (also it would have an explicit setEdgeId :) ? )

Reply:
Ok yes `SingleEdgeCursor` is good and avoids the confusion of calling it "iterator" or "explorer" :) Yes its main method would be `setEdgeId` (or `setEdge` or `moveToEdge(Id)` or `setToEdge(Id)?).

Reply:
> Probably the EdgeIterator is not a "cursor" because it has no direct setEdgeId method, only an implicit "next()" and so "iterator" is more appropriate (?)

Hppc does this:

```java
      for (IntCursor cursor : IntArrayList.from(3, 6, 5)) {
            System.out.println(cursor.value);
      }
```

and here we get the `IntCursor` when calling next on the `Iterator<IntCursor>`, so the `IntCursor` here is the equivalent to our `EdgeIteratorState` (the proxy object that gives access to the index and value of an int array for example).

Reply:
Ohhhh, yes yes yes yes yes, standard iteration syntax, please. :-)